### DISCUSSION TITLE
Moose-Input-Files: Handing Lists of Blocks

### DISCUSSION THREAD

**Discussion by jmeier** [idaholab/moose#26205] [Q&A General] [answered]
Dear Moose-Community,

I have a model with a large number of blocks. The geometry (nodes, elements, ...) is currently prepared outside of Moose in a GMSH file and this GMSH file is then imported into Moose using the FileMeshGenerator.

Properties etc. are assigned to the blocks in various commands in the Moose input file. To avoid having to do this individually for each block, I use variables that contain lists of block names.

I would like to maintain these lists outside of the Moose input file.

Question 1: Can additional lists of blocks be transferred to the Moose in a GMSH file? The assignment of several "PhysicalNames" per block is rejected with an error during import in Moose.

Question 2: Can, for example, a list of block names from a JSON file be read into a Moose variable using [JSONFileReader](https://mooseframework.inl.gov/source/userobjects/JSONFileReader.html) and then used? I haven't found anything about this in my search so far.

Question 3: Of course I can also have block lists written (automatically) to an *.i file and load this via "!import". Unfortunately, I get a message if I do not use *all* variables. Basically, I think this behaviour of warning of unused variables is very good. But how can I suppress the message for specific variables that they are not used?

**Comment by GiudGiud** [comment]
Hello

1) Not that I know of. The gmsh reader is in libmesh, feel free to add features there
2) I have never heard of anyone doing that. You absolutely could do it, but then you would need to modify the BlockRestrictableInterface to accept a list of blocks from JSON instead of from the input
3) You mean unused input parameters? Add `-w` on the command line to ignore that

Guillaume

**Reply by jmeier** [reply]
Dear Guillaume,

Thank you very much for the very quick reply.

Regarding 3: The switch "-w" converts the cancellation into a warning if there are any unused variables. However, I would like to completely deactivate this error/warning for some specific variables. Is it possible to create a fake user for a variable in the input file? Then it would not be unused from Moose's point of view and the usage of all other variables would continue to be monitored.

Jörg

**Reply by GiudGiud** [reply, marked as answer]
There's nothing in MOOSE but feel free to hack the input. 
For example, a, b and c are not used, but d is. 
your input
```
a = 3
b = 2
c = 34
d = ${fparse real_d_value + 0*a + 0*b + 0*c}
```
this wont complain that a,b and c are unused. But they wont matter.
You can do the same with strings by specifying a dummy object that is not used (a Function?) and give it some block lists you dont use elsewhere

**Reply by ABallisat** [reply]
Have you considered using a scripting language to automate the generation of the entire .i files? Using Python or similar makes this trivial, and if you are (perhaps) using the GMSH Python interface already it will be pretty easy to link together. That would give you the granularity of control you need for generating complex input files. I have found this makes handling input file that can run to thousands of lines much easier to wrangle and debug. I think you are hinting at this in question 3 but if you generate the whole input file then you can use some logic to choose which variables get added.

**Reply by GiudGiud** [reply]
There are various python codes for that.
see moose/python folder or see pygriffin (if you have access to that, never used it)

**Reply by jmeier** [reply]
Following your solution for variables, @GiudGiud, I wrote the following input file code snippet. The code is working inside my "primary" input file (the one I specify when calling the moose app).

```
# Code snippet for the definition of variables (integer, 
# double, string, array) and their use with "fake users".
# Even if these variables are not used elsewhere in the 
# input file, they do not generate an error message 
# for "unused variables".

fooInt = 7
fooDbl = 3.3
fooStr = "blubb bb 6"
fooArr = 'blubb blubb'

[Functions]
	
	[FakeUser_fooInt]
		type = ParsedFunction
		expression = 'a'
		symbol_names = 'a'
		symbol_values = '${fooInt}'
	[]
	
	[FakeUser_fooDbl]
		type = ParsedFunction
		expression = 'a'
		symbol_names = 'a'
		symbol_values = '${fooDbl}'
	[]
	
	[FakeUser_fooStr]
		type = ParsedFunction
		expression = 'a'
		symbol_names = 'a'
		symbol_values = '1'
		control_tags = ${fooStr}
	[]
	
	[FakeUser_fooArr]
		type = ParsedFunction
		expression = 'a'
		symbol_names = 'a'
		symbol_values = '1'
		control_tags = ${fooArr}
	[]

[]
```

But when I move the code to a "secondary" input file (one I include into my primary input file using `!include` I get the following error:
```
*** ERROR ***
/home/user/projects/moose_bare_tmech/models/tunnel-groups.i:12.3: no variable 'fooInt' found for substitution expression
```

I assume this behaviour is not intentional?

The motivation for my questions above is that I can have _parts_ of my input file created automatically (and then included via `!include).` For the creation of complete input files I probably need some more experience in Moose. Thanks for pointing out the Python codes @ABallisat and @GiudGiud.

**Reply by GiudGiud** [reply]
what's the order of includes?

```
fooInt = 7
fooDbl = 3.3
fooStr = "blubb bb 6"
fooArr = 'blubb blubb'

!include other_file_with_the_functions.i
```
?
@brandonlangley @joshuahansel this is something that got looked at recently right?

**Reply by jmeier** [reply]
I moved all of the code [above](https://github.com/idaholab/moose/discussions/26205#discussioncomment-7725725) (variables and fake-users) into the include. The idea is to integrate the externally automatically generated variables and fake users into my Moose model with one include command. 

However, I think I found the reason for the error cited [above](https://github.com/idaholab/moose/discussions/26205#discussioncomment-7725725): The "secondary" input file was saved as UTF8 (not ASCII). This leads to the effect that Moose prepends a space character in front of the variable name (e.g. ' fooInt' and not 'fooInt'). With this change, the include is now working. Thanks for your support!
